src/http/params.cr

```diff
@@ -78,6 +79,172 @@ module HTTP
       form_builder.to_s
     end
 
+    protected getter raw_params
+    def initialize(@raw_params)
+    end
+
+    def ==(other : self)
+      self.raw_params == other.raw_params
+    end
+
+    def ==(other)
+      false
+    end
+
+    # Returns first value for specified param name.
+    #
+    # ```
+    # params["email"]                # => "john@example.org"
+    # params["non_existent_param"]   # KeyError
+    # ```
+    def [](name)
+      raw_params[name].first
+    end
+
+    # Returns true if param with provided name exists.
+    #
+    # ```
+    # params.has_key?("email")       # => true
+    # params.has_key?("garbage")     # => false
+    # ```
+    delegate has_key?, raw_params
+
+    # Sets first value for specified param name.
+    #
+    # ```
+    # params["item"] = "pencil"
+    # ```
+    def []=(name, value)
+      raw_params[name] ||= [""]
+      raw_params[name][0] = value
+    end
+
+    # Returns all values for specified param name.
+    #
+    # ```
+    # params.fetch_all("item")       # => ["pencil", "book", "workbook"]
+    # ```
+    def fetch_all(name)
+      raw_params.fetch(name) { [] of String }
+    end
+
+    # Returns first value for specified param name.
+    #
+    # ```
+    # params.fetch("email")                # => "john@example.org"
+    # params.fetch("non_existent_param")   # KeyError
+    # ```
+    def fetch(name)
+      raw_params.fetch(name).first
+    end
+
+    # Returns first value for specified param name. Fallbacks to provided
+    # default value when there is no such param.
+    #
+    # ```
+    # params.fetch("email", "none@example.org")                # => "john@example.org"
+    # params.fetch("non_existent_param", "default value")      # => "default value"
+    # ```
+    def fetch(name, default)
+      raw_params.fetch(name, [default]).first
+    end
+
+    # Returns first value for specified param name. Fallbacks to return value
+    # of provided block.
+    #
+    # ```
+    # params.fetch("email") { raise InvalidUser("email is missing") }    # InvalidUser "email is missing"
+    # params.fetch("non_existent_param") { "default computed value" }    # => "default computed value"
+    # ```
+    def fetch(name, &block : -> String)
+      raw_params.fetch(name) { [block.call] }.first
```

Comment:
We could `yield` here. It's more code, but there's no captured block so there's no way a closure is created here. I guess something like `raw_params.fetch(name) { |key| yield key }.first`